### PR TITLE
Fixes #27074 - Migrate to @theforeman/vendor and @theforeman/vendor-dev pkg

### PR DESCRIPTION
**Pull request by sharvit**
Uses packages from:
https://github.com/sharvit/foreman-js

### CODE REVIEW

**Comment by Ron-Lavi** [comment]
What's the status of this PR? thanks :)

**Comment by sharvit** [comment]
> What's the status of this PR? thanks :)

I'm working on releasing a new version of `@theforeman/vendor` and then rebasing this PR.
Thanks for caring about vendor @Laviro :+1:

**Comment by sharvit** [comment]
1. dev-server is working well atm
2. production build and server working well atm
3. need to fix some small testing issues

**Comment by waldenraines** [comment]
I created a redmine issue for this - https://projects.theforeman.org/issues/27074

**Comment by sharvit** [comment]
# Status Update

I have just released `@theforeman/vendor` version `0.1.0-alpha.6` and updated this PR to use it.

You can test it by simply run:
```sh
rm -rf node_modules/
rm -rf package-lock.json
npm install

# running test should work
npm test

# running development server should work
foreman start

# building and running production should work
RAILS_ENV=production bundle exec rake assets:precompile
RAILS_ENV=production bundle exec rake webpack:compile
RAILS_ENV=production foreman start rails
```

## There is one issue when running `npm run lint`

Yesterday, `eslint-plugin-react` released version [7.14.0](https://github.com/yannickcr/eslint-plugin-react/releases/tag/v7.14.0) which break `patternfly-react-eslint`.
PR opend in `patternfly-react` to fix it:
https://github.com/patternfly/patternfly-react/pull/2334

**Comment by sharvit** [comment]
> ## There is one issue when running `npm run lint`
> Yesterday, `eslint-plugin-react` released version [7.14.0](https://github.com/yannickcr/eslint-plugin-react/releases/tag/v7.14.0) which break `patternfly-react-eslint`.
> PR opend in `patternfly-react` to fix it:
> [patternfly/patternfly-react#2334](https://github.com/patternfly/patternfly-react/pull/2334)

This is been fixed in `eslint-plugin-react` version [7.14.1](https://github.com/yannickcr/eslint-plugin-react/releases/tag/v7.14.1).
The PR in `patternfly-react` closed.

**Comment by sharvit** [comment]
https://github.com/theforeman/foreman/pull/6858 created to fix the linting issue

**Comment by sharvit** [comment]
It's green :horse_racing:

File: package.json
Comment (ezr-ondrej) [inline package.json]: Do we want to merge dependency on `alpha` version, or do we expect to get it at least beta as part of reviewing this? :slightly_smiling_face:

Comment (sharvit) [reply package.json]: Great question @ezr-ondrej :+1: 

IMO we should merge this PR with version `0.1.0` once it's ready and I got a strong feeling `0.1.0-alpha.6` will end up as `0.1.0`.

Do you have a different suggestion?

Comment (ohadlevy) [reply package.json]: does it really matter? IMHO we should merge when we think its ready, so we can find out the real issues afterwards? :)

File: script/npm_link_vendor.sh
Comment (ohadlevy) [inline script/npm_link_vendor.sh]: can this be configurable somehow? or at least print reasonable error messages regarding where it expects the directories to be at?

File: script/webpack-analyze
Comment (ohadlevy) [inline script/webpack-analyze]: why? this ensures that if webpack json output is not correct it will not run the analyzer.

File: webpack/assets/javascripts/react_app/components/AutoComplete/__tests__/integration.test.js
Comment (ohadlevy) [inline webpack/assets/javascripts/react_app/components/AutoComplete/__tests__/integration.test.js:8]: why can't we mock lodash? this seems like a lot of work for mocking? is this going to be common?

File: webpack/assets/javascripts/react_app/components/Editor/components/EditorView.js
Comment (ohadlevy) [inline webpack/assets/javascripts/react_app/components/Editor/components/EditorView.js:5]: just curios, how would be break the foreman-js bundle if we want to? it was shown to be faster in #6719 and wouldn't want to continue loading all vendors as one common chunk?

File: webpack/simple_named_modules.js
Comment (ohadlevy) [inline webpack/simple_named_modules.js]: ?

File: webpack/stories/index.js
Comment (ohadlevy) [inline webpack/stories/index.js]: won't this break storybook?

File: script/npm_link_vendor.sh
Comment (sharvit) [reply script/npm_link_vendor.sh]: Yes!

I will allow configuring via env variables and will add some documentations around it.

File: webpack/assets/javascripts/react_app/components/AutoComplete/__tests__/integration.test.js
Comment (sharvit) [reply webpack/assets/javascripts/react_app/components/AutoComplete/__tests__/integration.test.js:8]: We can mock lodash globally.
IMO we should rethink about how we do mocking.

Anyway, I would prefer it not to be apart of this PR if it's fine.

File: webpack/assets/javascripts/react_app/components/Editor/components/EditorView.js
Comment (sharvit) [reply webpack/assets/javascripts/react_app/components/Editor/components/EditorView.js:5]: I believe we can implement such a mechanism in `@theforeman/vendor` using the `customModules`:
https://github.com/theforeman/foreman-js/blob/master/packages/vendor-core/lib/modules.js#L64-L67

File: package.json
Comment (sharvit) [reply package.json]: > does it really matter? IMHO we should merge when we think its ready, so we can find out the real issues afterwards? :)

OK, agree :+1:

File: script/webpack-analyze
Comment (sharvit) [reply script/webpack-analyze]: My bad, leftover from some debugging session (it prevented me from seeing some logs).

File: webpack/simple_named_modules.js
Comment (sharvit) [reply webpack/simple_named_modules.js]: Leftover, `simple_named_modules.js` is deleted now as it is not needed anymore. (the vendor has its own way to simplify the names of the modules.

File: webpack/stories/index.js
Comment (sharvit) [reply webpack/stories/index.js]: No, and I have no explanation why it was there from the beginning as we load patternfly source file by es6 imports and let the webpack storybook to bundle them into one big bundle.

Comment (ohadlevy) [reply webpack/stories/index.js]: ah, thats easy, the application.js rails =require statements are not honored in storybook, so without it you dont have the relevant JS/CSS files...

Comment (sharvit) [reply webpack/stories/index.js]: Mmm, so I have tested it and it works without `patternfly/dist/js/patternfly`.
Maybe the code that uses the patternfly dist moved to webpack?

IMO we should not load the `app/assets/javascript/application.js` and the `webpack/assets/bundle.js` into the storybook and I plan to face it when implementing `@theforeman/env`.

**There is another issue that causes an error when running the storybook, I am uploading a fix as vendor alpha.7**

**Comment by sharvit** [comment]
# Update

1. I have released `@theforeman/vendor` version `0.1.0-alpha.7` with a fix for the storybook error and this PR now uses this version.
2. Added documentation to the storybook about using `@theforeman/vendor`.

**Comment by sharvit** [comment]
Test failures are not related.
The issue here: https://projects.theforeman.org/issues/27188

@ohadlevy can you revisit please?

File: .eslintrc
Comment (ohadlevy) [inline .eslintrc:5]: just wondering, shouldn't this end up being a plugin just like PF?

File: config/webpack.vendor.js
Comment (ohadlevy) [inline config/webpack.vendor.js:1]: why should we keep foreman vendor at all at this stage?

File: script/npm_link_foreman_js.sh
Comment (ohadlevy) [inline script/npm_link_foreman_js.sh:12]: wouldn't' ../ refer to the script path? imho you need to take into account the following

* add "set -x" # so it exits the script if something returned an error
* print the actual paths used, e.g. FOREMAN_JS_LOCATION is not defined, using <x> instead.
* check that <x> actually exists

Also, how would a user undo this action? a simple npm install will fix it?

**Comment by ohadlevy** [comment]
@sharvit the main concern i have is around how easy/hard would it be to contribute code now that you have to do it in multiple places, and I wonder if that everything that you added to vendor should really be there, for example:

* datatables? I wouldnt really want seeing anyone using it
* dashboard? would anyone want to build another dashboard? (perhaps openscap does..)
* anything that require code changes in foreman, maybe it shouldn't be in vendor, e.g. keep in vendor only packages that should be shared, not everything we used to keep in the original vendor?
* does it make sense to create a todo bundle? so hopefully it would reduced/removed over time (and not automatically push it to the new vendor?)

File: .eslintrc
Comment (sharvit) [reply .eslintrc:5]: Yes :)
I think it should be delivered with `@theforeman/env`

File: config/webpack.vendor.js
Comment (sharvit) [reply config/webpack.vendor.js:1]: Because I want the first version of the `vendor` to be a small increment of what we have in foreman atm.
Because the `intl` packages cause webpack to do dynamic imports and split into multiple files I would prefer to avoid it in the first version.

File: script/npm_link_foreman_js.sh
Comment (sharvit) [reply script/npm_link_foreman_js.sh:12]: @ohadlevy see the doc about using `foreman-vendor`:
https://github.com/theforeman/foreman/pull/6605/files#diff-c681f7a6d59f3d900f78ad1aea482096


> add "set -x" # so it exits the script if something returned an error
print the actual paths used, e.g. FOREMAN_JS_LOCATION is not defined, using instead.
check that actually exists

+1

> Also, how would a user undo this action? a simple npm install will fix it?

 Yes

**Comment by sharvit** [comment]
> * datatables? I wouldnt really want seeing anyone using it
> * dashboard? would anyone want to build another dashboard? (perhaps openscap does..)
> * anything that require code changes in foreman, maybe it shouldn't be in vendor, e.g. keep in vendor only packages that should be shared, not everything we used to keep in the original vendor?

My strategy here was to keep the vendor as close as I can to the current vendor so my rebases would be less difficult and so we can merge it asap.
I agree those packages are controversial but I would prefer doing those more radical changes after it is in use.

> * does it make sense to create a todo bundle? so hopefully it would reduced/removed over time (and not automatically push it to the new vendor?)

I think it makes more sense to me to add a depreciation mechanism to the `vendor`.

File: script/npm_link_foreman_js.sh
Comment (sharvit) [reply script/npm_link_foreman_js.sh:12]: Thanks @ohadlevy, updated

**Comment by ohadlevy** [comment]
@sharvit can you please have a look at the test failures? esp https://ci.theforeman.org/job/test_develop_pr_core/1260/database=postgresql-integrations,ruby=2.5,slave=fast/console ?

**Review by ohadlevy — COMMENTED** [review]
LGTM - @tbrisker @ekohl shall we give this a try and see what breaks?

**Comment by tbrisker** [comment]
Packaging for this (https://github.com/theforeman/foreman-packaging/pull/3828) isn't ready yet so everything will fail. 
From our last discussion, @ekohl was to talk to @evgeni and see which one of them finishes the packaging side by Wednesday (and @sharvit was supposed to update the discourse discussion with the summary so others are aware of it). Once that is done, we can merge both and see how badly things break.

**Comment by sharvit** [comment]
> @sharvit can you please have a look at the test failures? esp https://ci.theforeman.org/job/test_develop_pr_core/1260/database=postgresql-integrations,ruby=2.5,slave=fast/console ?

The error is `TypeError: Object.values is not a function`, `Object.values` is a NodeJS version >= 8 feature.
Can we upgrade NodeJS?

`Object.values` used by `@theforeman/vendor` so atm it support version >=8:
https://github.com/theforeman/foreman-js/blob/9106d175e4a5a7e914238e42c60a1a735788cfcd/packages/vendor/lib/Manifest.js#L20

It also means this is not correct and should be fixed:
https://github.com/theforeman/foreman-js/blob/master/packages/vendor/package.json#L12

Those are the options I can think about:

1. Upgrade to NodeJS 8 and update the `package.json` of `@theforeman/vendor` so it supports only NodeJS 8.

2. Update `@theforeman/vendor` to use a NodeJS 6 supported feature instead of the `Object.values`.

3. Update `@theforeman/vendor` to create a `build` process to the tools using `babel` so it will work with NodeJS version 6 while we can continue to use the latest features.

I would vote for option 1, If it is not possible atm I would prefer option 2 while we can implement option 3 at a future version.

**Comment by tbrisker** [comment]
@ekohl - iirc we only have node 6 in the builders (and test runners), not sure how much of an effort it would be to upgrade. If it isn't too difficult, it may be better to avoid incompatible code in the parts of the vendor that don't get built using babel.

**Comment by sharvit** [comment]
> @ekohl - iirc we only have node 6 in the builders (and test runners), not sure how much of an effort it would be to upgrade. If it isn't too difficult, it may be better to avoid incompatible code in the parts of the vendor that don't get built using babel.

Almost done migrating to `Object.keys` so it would be NodeJS 6 compatible :+1:

We would have to upgrade the `node` version since it is not supported anymore and tons of libraries dropping the NodeJS 6 support.

https://nodejs.org/en/about/releases/

**Comment by sharvit** [comment]
> > @ekohl - iirc we only have node 6 in the builders (and test runners), not sure how much of an effort it would be to upgrade. If it isn't too difficult, it may be better to avoid incompatible code in the parts of the vendor that don't get built using babel.
> 
> Almost done migrating to `Object.keys` so it would be NodeJS 6 compatible 
> 
> We would have to upgrade the `node` version since it is not supported anymore and tons of libraries dropping the NodeJS 6 support.
> 
> https://nodejs.org/en/about/releases/

Updated to use `@theforeman/vendor@0.1.0-alpha.8` which supports NodeJS version >= 6

**Comment by sharvit** [comment]
@tbrisker Does the current errors related to the `rpm` or it is something I should investigate and fix asap?
https://ci.theforeman.org/job/test_develop_pr_core/1279/database=postgresql-integrations,ruby=2.5,slave=fast/testReport/junit/(root)/AboutIntegrationTest/test_0001_about_page/

**Comment by tbrisker** [comment]
@sharvit looks like for some reason the integration tests don't manage to load the vendor modules, i don't think it's related to rpm since the tests run using the modules from npm

**Comment by sharvit** [comment]
@tbrisker can you retrigger a test?

**Comment by tbrisker** [comment]
@sharvit you can too! just comment with [test foreman] :wink:

**Comment by sharvit** [comment]
[test foreman]

**Comment by sharvit** [comment]
I was pretty sure I had to be a maintainer for that :/ thanks

**Comment by ekohl** [comment]
> @ekohl - iirc we only have node 6 in the builders (and test runners), not sure how much of an effort it would be to upgrade.

At this point it's quite an effort because it means moving to an SCL. All our RPM specs don't support that currently so we'd need to regenerate them all and rebuild all. If this vendor package works, it greatly reduces the number of packages so it may be easier. I'd like to avoid two massive migrations at the same time, especially in the very same area. Let's first stabilize this and then see about upgrading node itself.

**Comment by sharvit** [comment]
I'm confused, all the errors in Jenkins are disappearing when I'm using the chrome driver.

@ohadlevy @tbrisker @ekohl do you have any suggestion how to face with it?

**Comment by sharvit** [comment]
> I'm confused, all the errors in Jenkins are disappearing when I'm using the chrome driver.
> 
> @ohadlevy @tbrisker @ekohl do you have any suggestion how to face with it?

We decided to solve this issue by pushing forward https://github.com/theforeman/foreman/pull/6748
https://community.theforeman.org/t/rfc-changing-how-we-handle-webpack-building/11491/78?u=sharvit

**Comment by sharvit** [comment]
> We decided to solve this issue by pushing forward #6748
> https://community.theforeman.org/t/rfc-changing-how-we-handle-webpack-building/11491/78?u=sharvit

#6748 is merged now so I'm rebasing this PR and the ruby test error should be resolved.
Thanks @ezr-ondrej @tbrisker

**Comment by sharvit** [comment]
Looks like the failed test is not related?

**Comment by sharvit** [comment]
# Update

1. I have just released `@theforeman/vendor` version `0.1.0-alpha.9` and this PR use it now. 
    It fixes the issue that blocked the packaging: https://github.com/theforeman/foreman-packaging/pull/3882

2. This PR rebased and contains #6748 so the ruby test errors should be fixed now.

Ready for another session of reviews.

File: app/views/layouts/base.html.erb
Comment (tbrisker) [inline app/views/layouts/base.html.erb:14]: I think this change will prevent the `patternfly_and_overrides.scss` from the asset pipeline from overriding patternfly styling

File: config/webpack.config.js
Comment (tbrisker) [inline config/webpack.config.js]: I think we should keep this plugin so modules not provided by vendor are still imported using expectable and human readable names rather than numbers or hashes

File: jest.config.js
Comment (tbrisker) [inline jest.config.js]: Should this file live under webpack/ or config/?

File: config/webpack.config.js
Comment (tbrisker) [inline config/webpack.config.js:177]: Reading this plugin's description, i think it may have been the cause of the unexpected hash changes we were seeing in the past:

> Create smaller Lodash builds by replacing feature sets of modules with noop, identity, or simpler alternatives.

which is exactly what I discovered when trying to isolate the changes - modules that were being replaced by noop and identity.
Are we still using it in the vendor build? if so, I think we shouldn't - even though we may get a larger build, we want to make all of lodash available to consumers since we can't tell which modules aren't used.

File: script/npm_link_foreman_js.sh
Comment (tbrisker) [inline script/npm_link_foreman_js.sh:21]: shouldn't this be `"../node_modules/@theforeman"`?

File: script/npm_link_foreman_js.sh
Comment (tbrisker) [inline script/npm_link_foreman_js.sh:1]: Maybe add a comment here about what this script does?

File: webpack/assets/javascripts/bundle.js
Comment (tbrisker) [inline webpack/assets/javascripts/bundle.js]: Shouldn't this one also be moved to vendor?

File: config/webpack.config.js
Comment (sharvit) [reply config/webpack.config.js:177]: Agree, the new vendor **don't** use this plugin anymore so it contains the full `lodash` lib.
Once we stabilize the build, I'm open-minded trying more optimizations.

File: jest.config.js
Comment (sharvit) [reply jest.config.js]: +1 for `config` I will use that

File: script/npm_link_foreman_js.sh
Comment (sharvit) [reply script/npm_link_foreman_js.sh:21]: No, because you run this script from the root by using `npm run foreman-js:link`.
See `webpack/stories/docs/addingDependencies.md`

File: webpack/assets/javascripts/bundle.js
Comment (sharvit) [reply webpack/assets/javascripts/bundle.js]: https://github.com/theforeman/foreman-js/pull/14

Comment (sharvit) [reply webpack/assets/javascripts/bundle.js]: theforeman/foreman-js#14 is closed.
This PR updated to allow full access to lodash using the window object.

**Comment by sharvit** [comment]
@tbrisker can you revisit please?

File: script/npm_link_foreman_js.sh
Comment (tbrisker) [inline script/npm_link_foreman_js.sh:19]: this will make it `../../foreman-js` with the default, and add a `../` to the path given by the user which would be confusing

File: webpack/assets/javascripts/bundle.js
Comment (tbrisker) [inline webpack/assets/javascripts/bundle.js:29]: I may be missing it, where are we making window._ available if the bundle is removed?

Comment (sharvit) [reply webpack/assets/javascripts/bundle.js:29]: Lodash put it there automatically once you require it, the `bundle_lodash` just replaced the `window._` with another leaner object.

File: script/npm_link_foreman_js.sh
Comment (sharvit) [reply script/npm_link_foreman_js.sh:19]: imo it will make it easier because you will define `FOREMAN_JS_LOCATION` relatively to foreman root so you can have:
```
home/foreman
home/foreman-js
```

So inside foreman you can run:
```sh
FOREMAN_JS_LOCATION="../foreman-js" npm run foreman-js:lint
```

imo it will be confusing if you will need to set `FOREMAN_JS_LOCATION` relatively to the scripts folder.

Comment (tbrisker) [reply script/npm_link_foreman_js.sh:19]: didn't you reply below that this script runs from the foreman root because it is executed by npm? in that case `../../foreman-js` would be the wrong folder?

Comment (sharvit) [reply script/npm_link_foreman_js.sh:19]: I think there is a difference in how the `rm` and `ln` commands are working and their parameters.
You can try the script, I have just verified it is working.

Comment (sharvit) [reply script/npm_link_foreman_js.sh:19]: I think it is only the second parameter of the `ln` that need to be relative to root.

Comment (tbrisker) [reply script/npm_link_foreman_js.sh:19]: ah, i see, it's is because ln creates the link relative to the link location, not to the cwd. you could unite this line with the next but not critical.

**Review by tbrisker — DISMISSED** [review]
Thanks @sharvit ! code looks good, tested and everything seems to work well including several plugins both in development and production mode. This is ready to merge on my side once we figure out the packaging side.

**Comment by tbrisker** [comment]
@sharvit looks like there is a conflict in package.json, please rebase on latest develop

**Comment by sharvit** [comment]
> @sharvit looks like there is a conflict in package.json, please rebase on latest develop

This conflict will be solved only by updating the `foreman-js` project. (and releasing a new `@theforeman/vendor` to npm)

Notice there is another issue that should get fixed:
https://github.com/theforeman/foreman-packaging/pull/3882#issuecomment-510380473

**Comment by tbrisker** [comment]
@sharvit besides the packaging issue, this can't be merged right now because of a merge conflict with the current package.json in develop branch. rebasing should fix that.

**Comment by sharvit** [comment]
Thanks @tbrisker 

1. `@theforeman/vendor` version `0.1.0-alpha.10` has released.
    https://github.com/theforeman/foreman-js/compare/@theforeman/vendor@0.1.0-alpha.9...@theforeman/vendor@0.1.0-alpha.10
2. This PR updated to use the new vendor version.
3. The `babel` issue fixed.
4. This PR as rebased.

**Review by tbrisker — DISMISSED** [review]

**Comment by sharvit** [comment]
**Updates** 

1. `@theforeman/vendor` version `0.1.0-alpha.11` has released.
    https://github.com/theforeman/foreman-js/compare/@theforeman/vendor@0.1.0-alpha.10...@theforeman/vendor@0.1.0-alpha.11
2. This PR updated to use the new vendor version.
3. This PR updated with some scss cleanups

**Review by tbrisker — APPROVED** [review]
Thanks @sharvit !

**Pull request merged**
